lib/read.js

```diff
@@ -36,49 +36,55 @@ module.exports = read
  */
 
 function read (req, res, next, parse, debug, options) {
-  var length
-  var opts = options
-  var stream
-
   // read options
-  var encoding = opts.encoding !== null
-    ? opts.encoding
-    : null
-  var verify = opts.verify
-
-  try {
-    // get the content stream
-    stream = contentstream(req, debug, opts.inflate)
-    length = stream.length
-    stream.length = undefined
-  } catch (err) {
-    return next(err)
+  const charset = options.charset
+
+  // get the content stream
+  const contentEncoding = (req.headers['content-encoding'] || 'identity').toLowerCase()
+  debug('content-encoding "%s"', contentEncoding)
+
+  if (options.inflate === false && contentEncoding !== 'identity') {
+    return next(createError(415, 'content encoding unsupported', {
+      encoding: contentEncoding,
+      type: 'encoding.unsupported'
+    }))
   }
 
-  // set raw-body options
-  opts.length = length
-  opts.encoding = verify
-    ? null
-    : encoding
+  let stream
+  if (contentEncoding === 'identity') {
+    // set raw-body expected length
+    stream = req
+    options.length = req.headers['content-length']
```

Comment:
Improves the code by only accessing `req.headers['content-length']` and setting the `length` option for `raw-body` when the stream needs no decompression